File: ios/ios_app/SpeakingScanner/service/ble/BleService.swift
//
//  BleFacade.swift
//  SpeakingScanner
//
//  Created by Marc Plouhinec on 13/09/16.
//

import UIKit

///
/// Give access to the rest of the application for the BLE-to-WiFi bridge.
///
protocol BleService {
    
    ///
    /// Register a listener that is called when the application connection status via BLE to the ble-wifi-bridge changes.
    ///
    func registerBleConnectionStatusChangedListener(listener: @escaping (_ status: BleConnectionStatus) -> Void)
    
    ///
    /// Start searching for ble-wifi-bridge and handling communication.
    ///
    func start()
    
    ///
    /// Opposite of the start() function.
    ///
    func stop()
    
    ///
    /// Get the number of scanned documents.
    ///
    func getNbDocuments(onResponseReceived: @escaping (_ nbDocuments: Int, _ error: NSError?) -> Void)
    
    ///
    /// Get the names of the scanned documents by the given index and count.
    ///
    func getDocumentNamesByIndexAndCount(index: UInt16, count: UInt16, onResponseReceived: @escaping (_ documentNames: [String], _ error: NSError?) -> Void)
    
    ///
    /// Get the document preview by index.
    ///
    func getDocumentPreviewByIndex(index: UInt16, onResponseReceived: @escaping (_ documentPreview: Data, _ error: NSError?) -> Void, onProgressUpdate: @escaping (_ progressPercentage: Int) -> Void)
    
    ///
    /// Get the document in a compressed format by the given index.
    ///
    func getCompressedDocumentByIndex(index: UInt16, onResponseReceived: @escaping (_ documentData: Data, _ error: NSError?) -> Void, onProgressUpdate: @escaping (_ progressPercentage: Int) -> Void)
    
}

File: ios/wifiblebridge/service/bleService.js
const fs = require("fs");
const _ = require('lodash');
const configuration = require('../configuration');

/**
 * Provide documents to the BLE facade.
 *
 * @author Marc Plouhinec
 */
const bleService = {

    /**
     * Count the number of documents.
     *
     * @param {function(err: Error?, nbDocuments: Number)} callback Function called with the result.
     */
    getNbDocuments(callback) {
        fs.readdir(configuration.distributionReadyThumbnailsPath, (err, fileNames) => {
            if (err) {
                return callback(err, 0);
            }

            callback(null, fileNames.length);
        });
    },

    /**
     * Get the documents names by index and count.
     *
     * @param {number} documentIndex Index of the first document.
     * @param {number} documentCount Number of documents to return.
     * @param {function(err: Error?, documentNames: Array.<String>)} callback Function called with the result.
     */
    getDocumentNamesByIndexAndCount(documentIndex, documentCount, callback) {
        fs.readdir(configuration.distributionReadyThumbnailsPath, (err, fileNames) => {
            if (err) {
                return callback(err, []);
            }

            if (documentIndex < 0 || documentIndex >= fileNames.length || documentCount <= 0) {
                return callback(null, []);
            }

            const documentNames = _
                .chain(fileNames)
                .filter(name => name.toLowerCase().endsWith('.png'))
                .take(documentCount)
                .value();

            callback(null, documentNames);
        });
    },

    /**
     * Load in memory the preview picture of the document at the given index.
     *
     * @param {number} documentIndex Index of the document.
     * @param {function(err: Error?, data: Buffer)} callback Function called with the result.
     */
    getDocumentPreviewByIndex(documentIndex, callback) {
        this.getDocumentNamesByIndexAndCount(documentIndex, 1, (err, documentNames) => {
            if (err) {
                return callback(err, new Buffer(0));
            }

            if (documentNames.length === 0) {
                return new Buffer(0);
            }

            const filePath = configuration.distributionReadyThumbnailsPath + '/' + documentNames[0];
            fs.readFile(filePath, (err, data) => {
                if (err) {
                    return callback(err, new Buffer(0));
                }
                callback(null, data);
            });
        });
    },

    /**
     * Load in memory the picture of the document at the given index.
     *
     * @param {number} documentIndex Index of the document.
     * @param {function(err: Error?, data: Buffer)} callback Function called with the result.
     */
    getCompressedDocumentByIndex(documentIndex, callback) {
        this.getDocumentNamesByIndexAndCount(documentIndex, 1, (err, documentNames) => {
            if (err) {
                return callback(err, new Buffer(0));
            }

            if (documentNames.length === 0) {
                return new Buffer(0);
            }

            const filePath = configuration.distributionReadyPath + '/' + documentNames[0];
            fs.readFile(filePath, (err, data) => {
                if (err) {
                    return callback(err, new Buffer(0));
                }

                callback(null, data);
            });
        });
    }
};

module.exports = bleService;